src/box.jl

```diff
@@ -2,17 +2,19 @@ export Box
 
 type Box{N,D,L} <: Shape{N}
     c::SVector{N,Float64} # box center
-    p::SMatrix{N,N,Float64,L} # projection matrix to box coordinates
     r::SVector{N,Float64}   # "radius" (semi-axis) in each direction
+    p::SMatrix{N,N,Float64,L} # projection matrix to box coordinates
     data::D             # auxiliary data
+    (::Type{Box{N,D,L}}){N,D,L}(c,r,p,data) = new{N,D,L}(c,r,p,data)
```

Comment:
This is just an inner constructor.  I find that defining an inner constructor is a way to suppress automatic generation of the implicit outer constructor.  (Should've added some comments on this.  I will push an additional commit for this.)

I wanted to suppress this automatic generation of the implicit outer constructor, because its signature is very similar to our new type-stable outer constructor, and therefore users who intended to call the explicit outer constructor could call the auto-generated implicit outer constructor inadvertently.

For example, without the inner constructor, we have
```julia
julia> Box(SVector(0.0,0.0),SVector(1.0,2.0))
GeometryPrimitives.Box{2,Void,4}([0.0, 0.0], [1.0, 2.0], [1.0 0.0; 0.0 1.0], nothing)
```
This calls the implicit outer constructor, and therefore the second argument, `d = SVector(1.0,2.0)`, is directly stored to `Box.r` without being multiplied with `0.5`.

On the other hand, with the inner constructor, the implicit outer constructor is suppressed, and therefore the same code calls the explicit outer constructor:
```julia
julia> Box(SVector(0.0,0.0),SVector(1.0,2.0))
GeometryPrimitives.Box{2,Void,4}([0.0, 0.0], [0.5, 1.0], [1.0 0.0; 0.0 1.0], nothing)
```
where the second argument `d = SVector(1.0,2.0)` is correctly stored to `Box.r` after being multiplied with `0.5`.

Maybe this is not the best solution.  If you know a better approach, please let me know!